File: src/main/java/komu/blunt/types/TypeClass.java
package komu.blunt.types;

import com.google.common.base.Function;
import com.google.common.collect.Iterables;

import java.util.ArrayList;
import java.util.List;

public final class TypeClass {

    final List<String> superClasses;
    final List<ClassInstance> instances = new ArrayList<>();
    
    public TypeClass(List<String> superClasses) {
        this.superClasses = new ArrayList<>(superClasses);
    }

    public void addInstance(Qualified<Predicate> qual) {
        instances.add(new ClassInstance(qual));
    }
    
    public Iterable<Predicate> instancePredicates() {
        return Iterables.transform(instances, new Function<ClassInstance, Predicate>() {
            @Override
            public Predicate apply(ClassInstance instance) {
                return instance.qual.value;
            }
        });
    }
}

File: src/main/kotlin/komu/blunt/types/TypeClass.kt
package komu.blunt.types

import java.util.ArrayList

class TypeClass(val superClasses: List<String>) {

    val instances: MutableList<ClassInstance> = ArrayList<ClassInstance>()

    fun addInstance(qual: Qualified<Predicate>) {
        instances.add(ClassInstance(qual))
    }

    val instancePredicates: Collection<Predicate>
        get() = instances.map { it.qual.value }
}
